go/nbs/conjoiner.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package nbs
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/jpillora/backoff"
+)
+
+type conjoiner interface {
+	// ConjoinRequired tells the caller whether or not it's time to request a
+	// Conjoin, based upon the contents of |ts| and the conjoiner
+	// implementation's policy. Implementations must be goroutine-safe.
+	ConjoinRequired(ts tableSet) bool
+
+	// Conjoin uses |p| to conjoin some number of tables referenced by |mm|,
+	// allowing it to update |mm| with a new, smaller, set of tables that
+	// references precisely the same set of chunks. Before performing the
+	// conjoin, implementations should verify that a conjoin is actually
+	// currently needed; callers may be working with an out-of-date notion of
+	// upstream state. |novelCount|, the number of new tables the caller is
+	// trying to land, may be used in this determination. Implementations must
+	// be goroutine-safe.
+	Conjoin(mm manifest, p tablePersister, novelCount int, stats *Stats)
+}
+
+func newAsyncConjoiner(maxTables int) *asyncConjoiner {
+	return &asyncConjoiner{
+		waiters:   map[string]chan struct{}{},
+		maxTables: maxTables,
+	}
+}
+
+type asyncConjoiner struct {
+	mu        sync.RWMutex
+	waiters   map[string]chan struct{}
+	maxTables int
+}
+
+func (c *asyncConjoiner) ConjoinRequired(ts tableSet) bool {
+	return ts.Size() > c.maxTables
+}
+
+// Conjoin checks to see if there's already a conjoin underway for the store
+// described by |mm|. If so, it blocks until that conjoin completes. If not,
+// it starts one and blocks until it completes. Conjoin can be called
+// concurrently from many goroutines.
```

Comment:
Note here that Conjoin may fail (i.e. because of OOP actor), and callers must handle that.